File: LeetCode/101-200/106. 从中序与后序遍历序列构造二叉树.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
      int[] inorder;
    int[] postorder;
      HashMap<Integer,Integer> map = new HashMap<>();
    int postIndex;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        this.inorder = inorder;
        this.postorder = postorder;
        postIndex = inorder.length - 1;
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i],i);
        }
        return build(0, inorder.length - 1);
    }

    public TreeNode build(int left, int right) {
                System.out.println(left + "--" +right + "--" + postIndex);
        if (left > right) return null;
        int i = postorder[postIndex];
        TreeNode root = new TreeNode(i);
        Integer mid = map.get(i);
        postIndex--;
       root.right = build(mid + 1,right);
        root.left = build(left,mid - 1);
        return root;
    }
}

File: LeetCode/101-200/106. 从中序与后序遍历序列构造二叉树.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func buildTree(inorder []int, postorder []int) *TreeNode {
	idxMap := map[int]int{}
	for i, v := range inorder {
		idxMap[v] = i
	}
	var build func(int, int) *TreeNode
	build = func(inL, inR int) *TreeNode {
		if inR < inL {
			return nil
		}
		val := postorder[len(postorder)-1]
		postorder = postorder[:len(postorder)-1]
		root := &TreeNode{Val: val}
		ini := idxMap[val]
		root.Right = build(ini+1, inR)
		root.Left = build(inL, ini-1)
		return root
	}

	return build(0, len(postorder)-1)
}